### ISSUE TITLE
Github rate limit error WITH authentication

### ISSUE THREAD

**Issue by sergei-maertens** [jspm/github#97]
Hi, I'm not sure if this is on jspm's end or Github's, but in Travis I'm running into broken builds because of the dreaded Rate limit errors.

Before today, this was working perfectly. The error I'm seeing because of `jspm install` is:

    GitHub rate limit reached, with authentication enabled.

The build log is publicly visible: https://travis-ci.org/modelbrouwers/modelbrouwers/builds/144385431

As I said, I have set up the token before, and jspm correctly detects that auth is enabled. What gives?

**Comment by guybedford** [comment]
This is a regression unfortunately.

When we removed usernames from the github authentication process in https://github.com/jspm/github/pull/91 this stopped private authentication from working as private github needs the username and the token as the password.

It looks like we will need to revert the previous PR work, or merge in the ability to work with usernames again.

//cc @tamird

**Comment by tamird** [comment]
Am I missing something? I don't see that error in the build log at all.

> private github needs the username and the token as the password.

Now I'm completely lost. What does this have to do with private github? Everything here appears to be public.

**Comment by guybedford** [comment]
@tamird when configuring github credentials to use private repos, it is apparently necessary to use the username and token as the password in order for the authentication to work.

**Comment by guybedford** [comment]
And thanks for the quick response!

**Comment by tamird** [comment]
@guybedford how did you determine that? What does that have to do with this issue?

**Comment by guybedford** [comment]
@AdamBurgess reported this along with a PR at https://github.com/jspm/github/pull/95#issuecomment-232688931. This issue is exactly because the private repo authentication is failing, and thus hitting the rate limit.

**Comment by sergei-maertens** [comment]
For the record, all of this build should be public.

**Comment by guybedford** [comment]
@sergei-maertens is it a private github account though, even if public?

**Comment by guybedford** [comment]
(I may just be showing my ignorance here though)

**Comment by tamird** [comment]
This seems to have absolutely nothing to do with whatever is being discussed in #94.

@sergei-maertens I don't see any JSPM-related error in the log you linked.
@guybedford sounds to me like this error (wherever it occurred) was possibly a real case of hitting GH's API limit?

**Comment by sergei-maertens** [comment]
it's my personal github account, the very same that I'm using here to comment.

The build log in Travis seems to have changed because the build was triggered by something else, I'll see if I can dig it up somewhere again.

**Comment by sergei-maertens** [comment]
This one has the same issue: https://s3.amazonaws.com/archive.travis-ci.org/jobs/143824723/log.txt

Exerpt:

```
[0Ktravis_time:start:2547ba90
[0K$ jspm config registries.github.auth $GITHUBTOKEN

travis_time:end:2547ba90:start=1468224022243393123,finish=1468224022780825407,duration=537432284
[0Ktravis_fold:end:before_script.3
[0Ktravis_fold:start:before_script.4
[0Ktravis_time:start:08f59710
[0K$ jspm install
     Looking up [36mnpm:babel-core[39m
     Looking up [36mnpm:babel-runtime[39m
     Looking up [36mnpm:core-js[39m
     Looking up [36mnpm:URIjs[39m
     Looking up [36mnpm:aurelia-http-client[39m
     Looking up [36mnpm:aurelia-pal[39m
     Looking up [36mnpm:aurelia-pal-browser[39m
     Looking up [36mnpm:perfect-scrollbar[39m
     Looking up [36mnpm:q[39m
     Looking up [36mnpm:string.prototype.startswith[39m
     Creating registry cache...
     Looking up [36mgithub:components/jquery[39m
     Looking up [36mgithub:components/jqueryui[39m
     Looking up [36mgithub:twbs/bootstrap[39m
     Looking up [36mgithub:components/handlebars.js[39m
     Looking up [36mgithub:systemjs/plugin-json[39m
     Looking up [36mgithub:systemjs/plugin-css[39m
     Looking up [36mgithub:jspm/nodelibs-url[39m
     Looking up [36mgithub:twitter/typeahead.js[39m
     Looking up [36mgithub:sergei-maertens/django-ponyjs[39m
     Downloading [36mgithub:components/handlebars.js@2.0.0[39m
     Downloading [36mgithub:systemjs/plugin-json@0.1.0[39m
     Downloading [36mgithub:systemjs/plugin-css@0.1.13[39m
     Downloading [36mgithub:jspm/nodelibs-url@0.1.0[39m
     Downloading [36mgithub:sergei-maertens/django-ponyjs@0.1.2[39m
     Downloading [36mgithub:components/jquery@2.2.1[39m
     Downloading [36mgithub:components/jqueryui@1.11.4[39m
     Downloading [36mgithub:twbs/bootstrap@3.3.5[39m
     Downloading [36mgithub:twitter/typeahead.js@0.11.1[39m
     Looking up [36mnpm:url[39m
     Downloading [36mgithub:components/handlebars.js@3.0.3[39m
     Looking up [36mgithub:js-cookie/js-cookie[39m
     Downloading [36mgithub:js-cookie/js-cookie@2.1.0[39m

[33m[1mwarn [22m[39mError on download for [36mgithub:twbs/bootstrap[39m

     GitHub rate limit reached, with authentication enabled.
     The rate limit will reset in [36m5 minutes[39m.

[31m[1merr  [22m[39mError downloading [36mgithub:twbs/bootstrap[39m.

[33m[1mwarn [22m[39mInstallation changes not saved.

travis_time:end:08f59710:start=1468224022785126821,finish=1468224025060958172,duration=2275831351
[0K
[31;1mThe command "jspm install" failed and exited with 1 during .[0m

Your build has been stopped.
```

**Comment by tamird** [comment]
Have you confirmed that this is not a legit case of you having exceeded the API limit?

**Comment by sergei-maertens** [comment]
I could hardly imagine where, did no local jspm install and there were not
multiple builds running at the same time. I'm not sure where to start
diagnosing the problem, since I can't reproduce it locally either

**Comment by guybedford** [comment]
@sergei-maertens perhaps try alternative GitHub credentials and see if that helps?

**Comment by tamird** [comment]
Seems suspicious that you can't reproduce it locally, doesn't it?

**Comment by GregDracoulis** [comment]
I'm seeing this start to happen much more often in my Travis builds as well. Not sure if something has changed with JSPM, or GitHub has changed their rate limits (I'm thinking it's probably the former?)

@tamird this issue would be tricky to reproduce locally, since if you have already run `jspm install` you have your dependencies cached. You could clear your registry cache and remove the `jspm_packages` directory, then re-run `jspm install`. When you're running in a CI environment, you're always starting from a blank slate.

@guybedford I think there's just some confusion in terminology since the API is being used with GitHub account credentials to check out all `public_repo`s. I do believe that you still have to use usernames with the given access token to authenticate through the API?

**Comment by tamird** [comment]
Have you tried adding $HOME/.jspm to your Travis cache?

**Comment by GregDracoulis** [comment]
^ yeah, looks like adding that and/or `jspm_packages` to your `.travis.yml` like this would help:

```
cache:
  directories:
  - $HOME/.jspm
  - jspm_packages
```

I worry that this might be obscuring a deeper issue though. On my end, we can't use the Travis cache because we have an enterprise deployment without S3 set up to store the cache artifacts.

**Comment by sergei-maertens** [comment]
I have tried with a new token and adding the cache - however I can't get my deps installed in the first place, so the cache is doing nothing at this point.

Build log: https://travis-ci.org/modelbrouwers/modelbrouwers/builds/146775809

**Comment by guybedford** [comment]
@sergei-maertens is this on jspm 0.17 or jspm 0.16? If using jspm 0.16, try using jspm-github 0.13.13, rerunning the auth and see if that alters the rate limit behaviour. (this involves ensure jspm 0.16 is using jspm-github 0.13.13).

That will at least confirm if this is a regression or not.

**Comment by sergei-maertens** [comment]
This is 0.16.

After carefully picking the window to restart the build, it was successfull, and adding `$HOME/.jspm` to the Travis cache seems to work well enough.

I will try the suggestions later tonight (hopefully).

**Comment by guybedford** [comment]
This should be resolved in https://github.com/jspm/github/pull/112 now.

**Comment by guybedford** [comment]
Released in 0.13.17.
